active/InputValidation/InputValidation.md

```diff
@@ -0,0 +1,572 @@
+
+    public ObservableCollection<ValidationError> GetErrors(string propertyName)
+    {
+        return propertyName != null && Errors.ContainsKey(propertyName)
+            ? Errors[propertyName]
+            : new ObservableCollection<ValidationError>();
+    }
+
+    protected virtual void OnErrorsChanged(string propertyName)
+    {
+        ErrorsChanged?.Invoke(this, new DataErrorsChangedEventArgs(propertyName));
+    }
+
+    protected void ClearErrors()
+    {
+        foreach (var propertyName in Errors.Keys.ToList())
+        {
+            Errors.Remove(propertyName);
+            OnErrorsChanged(propertyName);
+        }
+    }
+}
+```
+#### Code – ValidationWrapper (no new APIs)
+ValidationWrapper derives from the above NotifyDataErrorInfoBase, and provides GetValue/SetValue methods for storage of subclass’ properties.  The SetValue performs validation.  This also provides an implementation of INotifyPropertyChanged.
+
+``` csharp
+public class ValidationWrapper : NotifyDataErrorInfoBase
+{
+
+    #region get/set
+    protected T GetValue<T>(string propertyName)
+    {
+        if (string.IsNullOrEmpty(propertyName))
+        {
+            throw new ArgumentException("Invalid property name", propertyName);
+        }
+
+        object value;
+        value = default(T);
+
+        return (T)value;
+    }
+
+    // Validate() gets called from inside SetValue
+    protected void SetValue<T>(string propertyName, T value)
+    {
+        if (string.IsNullOrEmpty(propertyName))
+        {
+            throw new ArgumentException("Invalid property name", propertyName);
+        }
+
+        var current = GetValue<T>(propertyName);
+
+        if (!value.Equals(current))
+        {
+            PropertyChanged?.Invoke(
+                    this, new PropertyChangedEventArgs(propertyName));
+            Validate(propertyName);
+        }
+    }
+    #endregion
+
+    #region Populate errors (C# - DataAnnotations)
+    private Task Validate(string propertyName)
+    {
+        ClearErrors();
+
+
+        return Task.Run(() =>
+        {
+            var val = GetPropertyValue(propertyName);
+            var results = new ObservableCollection<ValidationResult>();
+            var context = new ValidationContext(this, null, null) 
+                            { MemberName = propertyName };
+    
+            // Validator is from the .Net System.ComponentModel.DataAnnotations namespace
+            Validator.TryValidateProperty(val, context, results);
+
+            if (results.Any())
+            {
+                    foreach (var result in results)
+                    {
+                        Errors[propertyName].Add(new ValidationError(result.ErrorMessage));
+                    }
+            }
+
+                OnErrorsChanged(propertyName);
+            }
+
+        }
+    }
+    #endregion
+
+    #region privates
+    private object GetPropertyValue([CallerMemberName] string propertyName = "")
```

Comment:
Using `[CallerMemberName]` here is unnecessary as this si only called in one place and you'd never want the name of the caller passed.